# Análisis de datos con Pyhon - Pre-entrega
Nombre y apellido:

# Etapa 1

## Script básico
Actividades 2 y 3

In [1]:
# Generamos una lista de diccionarios que contenga algunos datos
ventas = [
  {'producto': 'notebook', 'precio': 1500.0, 'cantidad': 3},
  {'producto': 'monitor', 'precio': 500.0, 'cantidad': 5},
  {'producto': 'mouse', 'precio': 20.0, 'cantidad': 15}
]

In [11]:
# Calculamos las ventas acumuladas - map


In [9]:
# Calculamos las ventas acumuladas - listas por comprensión
venta_acumuladas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])
venta_acumuladas

7300.0

## Import Datasets
Actividad 1

In [4]:
import pandas as pd

In [5]:
# Import dataset ventas
df_ventas = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/ventas.csv")

In [6]:
# Import dataset clientes
df_clientes = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/clientes.csv")

In [7]:
# Import dataset marketing
df_marketing = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/marketing.csv")

## Dataframe ventas

### Análisis exploratorio de **datos**
Actividad 4

In [33]:
df_ventas


id_venta                producto   precio  cantidad fecha_venta  \
0          792       Cuadro decorativo   $69.94       5.0  02/01/2024   
1          811         Lámpara de mesa  $105.10       5.0  02/01/2024   
2         1156                Secadora   $97.96       3.0  02/01/2024   
3         1372                Heladera  $114.35       8.0  02/01/2024   
4         1546                Secadora  $106.21       4.0  02/01/2024   
...        ...                     ...      ...       ...         ...   
3030      1837         Horno eléctrico  $104.12       9.0  30/12/2024   
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

              categoria  
0            Decoración  
1            Decoración  
2     Electrodomésticos  
3     Electrodomésticos  
4     Electrodomésticos  
...                 ...  
3030  Electrodomésticos  
3031        Electrónica  
3032        Electrónica  
3033        Electrónica  
3034        Electrónica  

[3035 rows x 6 columns]

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [22]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [23]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head()

id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos

In [25]:
# Usar sample() para tomar registros al azar
df_ventas.sample(10)

id_venta               producto   precio  cantidad fecha_venta  \
2094      2464                 Tablet   $69.17       9.0  01/09/2024   
2864        67  Elementos de cerámica  $109.07      12.0  10/12/2024   
1459      1662             Microondas   $33.73       1.0  12/06/2024   
2419      2352              Televisor   $82.27       2.0  12/10/2024   
652       1099               Batidora   $39.73      11.0  16/03/2024   
1445       637  Elementos de cerámica   $44.23       4.0  11/06/2024   

              categoria  
2094        Electrónica  
2864         Decoración  
1459  Electrodomésticos  
2419        Electrónica  
652   Electrodomésticos  
1445         Decoración

In [26]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].sample(10)

844      $41.72
124      $40.97
972      $29.68
928     $118.03
2760    $104.95
1805    $123.50
787     $103.69
801      $64.18
2242    $120.13
581      $86.29
Name: precio, dtype: object

In [27]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(10)

397      8.0
310      8.0
219      9.0
109     10.0
1204     9.0
1744     2.0
1449    10.0
389      1.0
1350     2.0
1803     3.0
Name: cantidad, dtype: float64

In [28]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"].sample(10)

90      12/01/2024
292     03/02/2024
2182    13/09/2024
2779    29/11/2024
2517    27/10/2024
326     07/02/2024
2882    12/12/2024
2033    25/08/2024
746     26/03/2024
1027    24/04/2024
Name: fecha_venta, dtype: object

In [29]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

categoria
Decoración           1015
Electrónica          1012
Electrodomésticos    1008
Name: count, dtype: int64

### Calidad de datos
Actividad 5

In [34]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [35]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="producto")

id_venta         producto   precio  cantidad fecha_venta    categoria
845       300  Adorno de pared   $67.75       9.0  06/04/2024   Decoración
833       300  Adorno de pared   $67.75       9.0  06/04/2024   Decoración
857       345         Alfombra   $56.78       9.0  07/04/2024   Decoración
832       145         Alfombra   $41.72       8.0  06/04/2024   Decoración
868       345         Alfombra   $56.78       9.0  07/04/2024   Decoración
..        ...              ...      ...       ...         ...          ...
854      2108       SmartWatch   $79.37      10.0  06/04/2024  Electrónica
831      2778       SmartWatch   $70.66       7.0  05/04/2024  Electrónica
825      2778       SmartWatch   $70.66       7.0  05/04/2024  Electrónica
877      2373       Smartphone  $123.60      12.0  07/04/2024  Electrónica
866      2373       Smartphone  $123.60      12.0  07/04/2024  Electrónica

[70 rows x 6 columns]

In [36]:
# Buscamos datos nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [37]:
# Visualizar los nulos
# df_ventas[df_ventas.isnull().any(axis=1)]
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [65]:
# Analizar los precios de los NaN a ver que opciones tenemos
#df_ventas[df_ventas["producto"]=="Elementos de cerámica"]
df_ventas["producto"].value_counts()
# df_ventas["producto"].unique()


producto
Lámpara de mesa           176
Auriculares               143
Microondas                135
Cafetera                  117
Smartphone                101
Rincón de plantas         101
Adorno de pared           100
Aspiradora                100
Cortinas                  100
Freidora eléctrica        100
Secadora                  100
Horno eléctrico           100
Plancha de vapor          100
Cuadro decorativo         100
Televisor                 100
Alfombra                  100
SmartWatch                100
Batidora                  100
Cámara digital            100
Lavadora                  100
Espejo decorativo         100
Jarrón decorativo         100
Laptop                    100
Consola de videojuegos     99
Elementos de cerámica      99
Proyector                  99
Parlantes Bluetooth        98
Tablet                     58
Heladera                   48
Candelabro                 24
Name: count, dtype: int64

### Observaciones




* Cantidad de registros: 3035
* Columnas 6
* 35 duplicados, eliminarlos
* Algunos nulos (2), 1% del faltante, dropna
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64
* Campo fecha: ajustar a dtype datetime

## Dataframe marketing

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [45]:
df_marketing

id_campanha            producto  canal  costo fecha_inicio   fecha_fin
0            74     Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1            12              Tablet   RRSS   3.40   26/03/2024  13/05/2024
2            32     Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3            21          Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4            58            Alfombra  Email   4.25   31/03/2024  05/05/2024
..          ...                 ...    ...    ...          ...         ...
85           70          Aspiradora     TV   3.06   13/12/2024  29/12/2024
86           89           Televisor     TV   4.98   13/12/2024    8/2/2025
87           68   Rincón de plantas     TV   5.81   17/12/2024   14/2/2025
88           33            Secadora  Email   3.80   20/12/2024    7/1/2025
89           11  Freidora eléctrica   RRSS   5.27   29/12/2024   21/1/2025

[90 rows x 6 columns]

In [39]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


In [40]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_marketing.head()

id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024

In [46]:
# Analizamos el campo costo
df_marketing["costo"].sample(10)

21    5.88
42    5.57
47    2.95
35    6.06
51    6.47
33    4.38
25    4.52
88    3.80
37    5.46
64    4.25
Name: costo, dtype: float64

In [47]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(10)

32    08/07/2024
78    05/11/2024
57    28/08/2024
15    12/05/2024
17    22/05/2024
85    13/12/2024
88    20/12/2024
64    25/09/2024
73    31/10/2024
59    09/09/2024
Name: fecha_inicio, dtype: object

In [48]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(10)

36    03/08/2024
87     14/2/2025
71    24/11/2024
21    29/06/2024
4     05/05/2024
46    25/08/2024
34    03/09/2024
68    16/11/2024
14    13/06/2024
18    02/07/2024
Name: fecha_fin, dtype: object

### Calidad de los datos


In [84]:
# Buscamos registros duplicados
df_marketing.duplicated().sum()

np.int64(0)

In [85]:
# Buscamos datos nulos
df_marketing.isnull().sum()


id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

### Observaciones


Cantidad de registros: 90
Columnas 6
0 duplicados
0 nulos


## Dataframe clientes

### Analisis exploratorio de datos

In [86]:
df_clientes

id_cliente               nombre  edad         ciudad  ingresos
0             1      Aloysia Screase    44  Mar del Plata  42294.68
1             2  Kristina Scaplehorn    25        Posadas  24735.04
2             3       Filip Castagne    50    Resistencia  35744.85
3             4          Liuka Luard    39   Bahía Blanca  27647.96
4             5        Dore Cockshtt    28        Rosario  28245.65
..          ...                  ...   ...            ...       ...
562         563        Dione Forsyde    29        Posadas  26757.73
563         564          Fleming Gow    39       Santa Fe  43674.96
564         565      Jewelle Mabbett    33        Córdoba  30522.64
565         566          Lauri Munns    23    Resistencia  31259.14
566         567          Micah Matis    31     Corrientes  42927.86

[567 rows x 5 columns]

In [87]:
# Aplicamos info() para familiarizarnos con el dataframe
df_clientes.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 22.3+ KB


In [88]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_clientes.head()


id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25        Posadas  24735.04
2           3       Filip Castagne    50    Resistencia  35744.85
3           4          Liuka Luard    39   Bahía Blanca  27647.96
4           5        Dore Cockshtt    28        Rosario  28245.65

In [89]:
# Usar sample() para tomar registros al azar
df_clientes.sample(10)


id_cliente               nombre  edad         ciudad  ingresos
364         365  Ellwood Nancekivell    38        Posadas  36909.15
482         483       Belinda Kenner    41        Posadas  35624.46
555         556           Bram Corps    48       Santa Fe  28839.22
65           66   Mohandas Pickworth    34  Mar del Plata  27487.51
206         207      Lissi Simondson    48        Rosario  39751.41
48           49       Dyanna Jimpson    51     Corrientes  39132.63
557         558   Ulrikaumeko Silber    52    Resistencia  21856.55
127         128          Lacie Cline    81          Salta  16595.30
15           16       Garrott Bogart    32        Rosario  16480.55
448         449     Tadeas Millwater    35     Corrientes  39688.56

In [90]:
# Analizamos el campo ciudad
# Si es numericos se puede aplicar describe()
df_clientes["ciudad"].sample(10)


433           Merlo
77            Salta
98         Santa Fe
81          Rosario
112      Corrientes
183    Bahía Blanca
132         Córdoba
485     Resistencia
82          Rosario
161        Santa Fe
Name: ciudad, dtype: object

In [93]:
# Analizamos el campo edad
# si es numerico se puede aplicar describe()
df_clientes["edad"].sample(10)


111    26
83     36
288    40
244    22
491    31
58     45
439    41
305    40
184    46
509    51
Name: edad, dtype: int64

# Calidad de los datos

In [95]:
# Buscamos registros duplicados
df_clientes.duplicated().sum()


np.int64(0)

In [94]:
# Buscamos datos nulos
df_clientes.isnull().sum()


id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

### Observaciones

Cantidad de registros: 567
Columnas 5
0 duplicados
0 nulos

# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas

* 35 duplicados, eliminarlos - se eliminaron 35 duplicados
* Algunos nulos (2), 1% del faltante, dropna - se eliminaron 2 registros con 2 NaN
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64 - Ok
* Campo fecha: ajustar a dtype datetime - Ok

In [51]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [52]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

np.int64(0)

In [58]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [68]:
# Validamos que no haya nulos
#df_ventas.isnull().sum()
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2998 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     2998 non-null   int64         
 1   producto     2998 non-null   object        
 2   precio       2998 non-null   float64       
 3   cantidad     2998 non-null   int64         
 4   fecha_venta  2998 non-null   datetime64[ns]
 5   categoria    2998 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 164.0+ KB


In [83]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$","")

AttributeError: Can only use .str accessor with string values!

In [74]:
# Validamos que precio no contenga $
df_ventas["precio"].sample(3)

1590    82.56
991     73.95
1180    27.86
Name: precio, dtype: float64

In [75]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype(float)

In [76]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["cantidad"]= df_ventas["cantidad"].astype(int)

In [77]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes

dtype('int64')

In [78]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [79]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

dtype('<M8[ns]')

### Dataframe marketing

Convertir a datetime fecha_inicio / fecha_fin

In [ ]:
df_marketing.info()

In [80]:
# Convertir el campo fecha_inicio a dtype datetime
df_marketing['fecha_inicio'] = pd.to_datetime(df_marketing['fecha_inicio'], format="%d/%m/%Y")

In [81]:
# Convertir el campo fecha_fin a dtype datetime
df_marketing['fecha_fin'] = pd.to_datetime(df_marketing['fecha_fin'], format="%d/%m/%Y")

In [82]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas
df_ventas["valor_venta"] = df_ventas["precio"] * df_ventas["cantidad"]

In [ ]:
# Visualizamos
df_ventas.head(2)

In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo
tabla_ventas = (df_ventas.groupby("producto", as_index=False)["valor_venta"].sum())
tabla_ventas.head()

In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante
tabla_ventas.describe()

In [ ]:
# Agregamos un plot para visualizar graficamente la distribución
import seaborn as sns
import matplotlib.pyplot as plt

sns.histplot(tabla_ventas["valor_venta"], bins=30, kde=True, edgecolor="black")
plt.title("Distribución de Valor de Venta con Curva de Densidad")
plt.xlabel("Valor de venta")
plt.ylabel("Frecuencia")
plt.show()

In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)
umbral_ventas = tabla_ventas["valor_venta"].quantile(0.75)
umbral_ventas

In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")
productos_top = tabla_ventas[tabla_ventas["valor_venta"] >= umbral_ventas]
productos_top

## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
df_ventas.columns

In [ ]:
# Agrupamos por categoria y agregamos la suma
ventas_categoria = df_ventas.groupby("categoria", as_index=False)["valor_venta"].sum()
ventas_categoria.head()

In [ ]:
ventas_categoria

In [ ]:
# Ordenar y mostrar los resultados
ventas_categoria

## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
df_ventas.info()

In [ ]:
df_marketing.info()

In [ ]:
# Vemos la cantidad de productos unicos en ventas
df_ventas["producto"].nunique()

In [ ]:
# Vemos la cantidad de productos unicos en marketing
df_marketing["producto"].nunique()

In [ ]:
# Listamos los registros de ventas ordenados por producto ascendiente
df_ventas.sort_values(by="producto", ascending=True )

In [ ]:
# Listamos los registros de marketing ordenados por producto ascendiente
df_marketing.sort_values(by=["producto","canal"])

In [ ]:
# Para evitar duplicados, habria que agrupar marketing antes de hacer el merge (a costa de perder la fecha)
df_marketing_avg = df_marketing.groupby(["producto"], as_index=False)["costo"].mean()
df_marketing_avg

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'
# Aqui tienen dos caminos, unir ventas con df_marketing que tiene duplicados
ventas_marketing = pd.merge(df_ventas, df_marketing, on="producto", how="left")
# O, unir ventar con df_marketing_avg que no tiene duplicados, pero no tiene fechas
# ventas_marketing = pd.merge(df_ventas, df_marketing_avg, on="producto", how="inner")

In [ ]:
# Verificar coincidencias (muestra)
ventas_marketing.sort_values(by=["producto", "precio"], ascending=True)

# Nota1: si unieron ventas con marketing, obtienen 8994 registros
# Nota2: si unieron ventas con marketing_avg obtienen 2998 registros

### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Nota: si trabajan con el datafrane marketing pueden aplicar el filtro,
# si trabajan con dataframe marketing_avg, al no tener fecha, no apliquen el filtro.

ganancias_marketing = ventas_marketing.query("fecha_venta > fecha_inicio & fecha_venta < fecha_fin")

# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"
ganancias_marketing = ganancias_marketing[["producto", "valor_venta", "costo"]]
ganancias_marketing.sort_values(by="producto", ascending=True)

In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"
ganancias_marketing = ganancias_marketing.groupby("producto")[["valor_venta","costo"]].sum()

In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
ganancias_marketing["ganancia_neta"] = ganancias_marketing['valor_venta']-ganancias_marketing['costo']
ganancias_marketing=ganancias_marketing.sort_values(by="ganancia_neta", ascending=False)

In [ ]:
# Crear un gráfico de barras apilado para comparar valor_venta y costo por producto
ganancias_marketing[["valor_venta", "costo"]].plot(kind='bar', stacked=True, figsize=(12, 6))
plt.title("Comparación de Valor de Venta y Costo por Producto")
plt.xlabel("Producto")
plt.ylabel("Monto")
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()

# Solución al problema de los duplicados

In [ ]:
# Hacemos el merge entre marketing y ventas por columna producto
marketing_ventas = pd.merge(df_marketing, df_ventas, on="producto", how="inner")
marketing_ventas.sort_values(by="producto", ascending=True)

In [ ]:
# filtramos solo aquellos productos que se hayan vendido dentro de la campaña
marketing_ventas_pp1 = marketing_ventas.query("fecha_venta > fecha_inicio & fecha_venta < fecha_fin")

In [ ]:
# Calculamos el total_ventas por producto haciendo valor_venta * cantidad
marketing_ventas_pp2 = marketing_ventas_pp1.copy() # para eliminar la advertencia
marketing_ventas_pp2["total_venta"] = (
    marketing_ventas_pp2["valor_venta"] * marketing_ventas_pp2["cantidad"]
)


In [ ]:
# Agrupamos el dataframe por producto usando .agg
# Agregar total_venta como el promedio, elimina el impacto de los datos duplicados
# Y seguimos agegando costo como la suma de lo invertido en cada canal
ganancias_marketing = marketing_ventas_pp2.groupby("producto").agg(
    {
        "total_venta": "mean",
        "costo": "sum",
    }
)

In [ ]:
# A partir de venta_total y costo, calculamos la ganancia obtenida
ganancias_marketing["ganancia"]= ganancias_marketing["total_venta"]-ganancias_marketing["costo"]
ganancias_marketing = ganancias_marketing.sort_values(by="ganancia", ascending=False)
ganancias_marketing.head(3)

In [ ]:
# Crear un gráfico de barras apilado para comparar valor_venta y costo por producto
ganancias_marketing[["total_venta", "costo"]].plot(kind='bar', stacked=True, figsize=(12, 6))
plt.title("Comparación de Valor de Venta y Costo por Producto")
plt.xlabel("Producto")
plt.ylabel("Monto")
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()

Ahora en el gráfico los costos de campañas de marketing sin más signficativos.